cmd/maestro/server/grpc_server.go

```diff
@@ -176,110 +203,118 @@ func (svr *GRPCServer) Subscribe(subReq *pbv1.SubscriptionRequest, subServer pbv
+	manifest, err := api.CloudEventToJSONMap(evt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert cloudevent to resource manifest: %v", err)
+	}
+	resource.Manifest = manifest
+
+	switch eventDataType {
+	case workpayload.ManifestEventDataType:
+		resource.Type = api.ResourceTypeSingle
+	case workpayload.ManifestBundleEventDataType:
+		resource.Type = api.ResourceTypeBundle
+	default:
+		return nil, fmt.Errorf("unsupported cloudevents data type %s", eventDataType)
+	}
+
+	return resource, nil
 }
 
-func encode(resource *api.Resource) (*cloudevents.Event, error) {
-	eventType := types.CloudEventsType{
-		CloudEventsDataType: payload.ManifestEventDataType,
-		SubResource:         types.SubResourceStatus,
-		Action:              config.UpdateRequestAction,
-	}
+// encode translates a resource to a cloudevent
+func encode(resource *api.Resource) (*ce.Event, error) {
+	return api.JSONMAPToCloudEvent(resource.Status)
+}
 
-	evt := types.NewEventBuilder(resource.Source, eventType).
-		WithResourceID(resource.ID).
-		WithResourceVersion(int64(resource.Version)).
-		WithClusterName(resource.ConsumerID).
-		NewEvent()
+// respondResyncStatusRequest responds to the status resync request by comparing the status hash of the resources
+// from the database and the status hash in the request, and then respond the resources whose status is changed.
+func (svr *GRPCServer) respondResyncStatusRequest(ctx context.Context, evt *ce.Event) error {
+	objs, serviceErr := svr.resourceService.FindBySource(ctx, evt.Source())
+	if serviceErr != nil {
+		return fmt.Errorf("failed to list resources: %s", serviceErr)
+	}
 
-	resourceStatusJSON, err := json.Marshal(resource.Status)
+	statusHashes, err := payload.DecodeStatusResyncRequest(*evt)
 	if err != nil {
-		return nil, fmt.Errorf("failed to marshal resource status, %v", err)
-	}
-	resourceStatus := &api.ResourceStatus{}
-	if err := json.Unmarshal(resourceStatusJSON, resourceStatus); err != nil {
-		return nil, fmt.Errorf("failed to unmarshal resource status, %v", err)
+		return fmt.Errorf("failed to decode status resync request: %v", err)
 	}
 
-	contentStatusJSON, err := json.Marshal(resourceStatus.ContentStatus)
-	if err != nil {
-		return nil, fmt.Errorf("failed to marshal content status, %v", err)
+	if len(statusHashes.Hashes) == 0 {
+		// publish all resources status
+		for _, obj := range objs {
+			svr.eventBroadcaster.Broadcast(obj)
+		}
+
+		return nil
 	}
-	contentStatusJSONStr := string(contentStatusJSON)
-
-	statusPayload := &payload.ManifestStatus{
-		Conditions: resourceStatus.ReconcileStatus.Conditions,
-		Status: &workv1.ManifestCondition{
-			Conditions: resourceStatus.ReconcileStatus.Conditions,
-			StatusFeedbacks: workv1.StatusFeedbackResult{
-				Values: []workv1.FeedbackValue{
-					{
-						Name: "status",
-						Value: workv1.FieldValue{
-							Type:    workv1.JsonRaw,
-							JsonRaw: &contentStatusJSONStr,
-						},
-					},
-				},
-			},
-		},
+
+	for _, obj := range objs {
```

Comment:
check/filter event type matching?

Reply:
do you mean `single` or `bundle` match?

Reply:
yes, you can get the event data type from the resync request event, then you can only response the type matched, right?

Reply:
I think source is will be sufficient to filter the resources.
For hub controller, the matched type will always be `bundle`, and using source ID will filter all the resources from that source.
But for general grpc client, the resource type will be unknown, it can be `single` or `bundle`, so we can't filter the resource with type.
make sense?

Reply:
thank! done.